### ISSUE TITLE
DB write conflicts

### ISSUE THREAD

**Issue by houjieth** [prysmaticlabs/prysm#514]
Based on my experiment, I found that when saving attestation hash lists with a DB key that contains suffix as a blockhash, it actually ruins the storage for that block.

reading block with db key: [98 108 111 99 107 45 **144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

got block with slot 4, hash [**144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

Saved attestation hash list with db key: [97 116 116 101 115 116 97 116 105 111 110 72 97 115 104 101 115 45 **144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

reading block with db key: [98 108 111 99 107 45 **144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

got block with slot 0, hash [177 96 222 67 175 80 135 91 174 86 199 237 141 71 75 205 66 189 212 180 68 136 185 134 237 82 129 27 247 25 228 18]
(?????)

If we change attestationHashesPrefix into a suffix (keep string content the same), we get

reading block with db key: [98 108 111 99 107 45 **144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

got block with slot 4, hash [**144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

Saved attestation with db key: [**144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75** 97 116 116 101 115 116 97 116 105 111 110 72 97 115 104 101 115 45]

reading block with db key: [98 108 111 99 107 45 **144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]

got block with slot 4, hash [**144 122 112 59 83 227 174 173 221 76 194 9 118 123 241 72 121 146 240 99 96 135 201 73 131 190 224 230 252 1 126 75**]
(works as expected)

I found this when working on the unit test `TestProcessingBlocks`. Any ideas for this weird behavior?

**Comment by terenc3t** [comment]
Interesting. It's something to watch out for as #503 progresses

**Comment by rawfalafel** [comment]
Hey @houjieth , could you show an example of a failing test case?

**Comment by rawfalafel** [comment]
It's hard to tell without reproducible code, but it sounds like you were able to work around the issue by changing the key to a suffix. Is it okay if I close this issue?

**Comment by rawfalafel** [comment]
I looked into this a little more, and there's a bug/limitation in the in memory database. Here's a failing test case:

```go
func TestDB(t *testing.T) {
	b, db := startInMemoryBeaconChain(t)
	defer db.Close()

	hash := [32]byte{}
	key1 := blockKey(hash)
	b.db.Put(key1, []byte{'A'})

	v, _ := b.db.Get(key1)
	if v[0] != 'A' {
		t.Fatalf("initial write failed: %x", v)
	}

	key2 := attestationHashListKey(hash)
	b.db.Put(key2, []byte{'B'})

	v, _ = b.db.Get(key1)
	if v[0] != 'A' {
		t.Fatalf("initial write overwritten: %x", v)
	}
}
```

[Here's the line where things go wrong](https://github.com/prysmaticlabs/prysm/blob/master/shared/database/inmemory.go#L46). The inmemory database uses 32 byte arrays as the key, and truncates from the right if the provided key is too long. This explains why your issue was resolved by making the prefix a suffix, and the test would have succeeded if it wasn't an inmemory database. 

We could fix this, but #503 will deprecate the in memory database anyways, so I'm going to close it.

Thanks for creating this ticket, I'm surprised it took so long for someone to run into this.
